copycat-api/src/main/java/org/apache/kafka/copycat/connector/Connector.java

```diff
@@ -0,0 +1,121 @@
+ * Connectors manage integration of Copycat with another system, either as an input that ingests
+ * data into Kafka or an output that passes data to an external system. Implementations should
+ * not use this class directly; they should inherit from SourceConnector or SinkConnector.
+ * </p>
+ * <p>
+ * Connectors have two primary tasks. First, given some configuration, they are responsible for
+ * creating configurations for a set of {@link Task}s that split up the data processing. For
+ * example, a database Connector might create Tasks by dividing the set of tables evenly among
+ * tasks. Second, they are responsible for monitoring inputs for changes that require
+ * reconfiguration and notifying the Copycat runtime via the ConnectorContext. Continuing the
+ * previous example, the connector might periodically check for new tables and notify Copycat of
+ * additions and deletions. Copycat will then request new configurations and update the running
+ * Tasks.
+ * </p>
+ */
+public abstract class Connector {
+
+    protected ConnectorContext context;
+
+    /**
+     * Initialize this connector, using the provided ConnectorContext to notify the runtime of
+     * input configuration changes.
+     * @param ctx context object used to interact with the Copycat runtime
+     */
+    public void initialize(ConnectorContext ctx) {
+        context = ctx;
+    }
+
+    /**
+     * <p>
+     * Initialize this connector, using the provided ConnectorContext to notify the runtime of
+     * input configuration changes and using the provided set of Task configurations.
+     * This version is only used to recover from failures.
+     * </p>
+     * <p>
+     * The default implementation ignores the provided simply ignores the provided Task
+     * configurations. During recovery, Copycat will request an updated set of configurations and
+     * update the running Tasks appropriately. However, Connectors should implement special
+     * handling of this case if it will avoid unnecessary changes to running Tasks.
+     * </p>
+     *
+     * @param ctx context object used to interact with the Copycat runtime
+     * @param taskConfigs existing task configurations, which may be used when generating new task configs to avoid
+     *                    churn in partition to task assignments
+     */
+    public void initialize(ConnectorContext ctx, List<Properties> taskConfigs) {
+        context = ctx;
+        // Ignore taskConfigs. May result in more churn of tasks during recovery if updated configs
+        // are very different, but reduces the difficulty of implementing a Connector
+    }
+
+    /**
+     * Start this Connector. This method will only be called on a clean Connector, i.e. it has
+     * either just been instantiated and initialized or {@link #stop()} has been invoked.
+     *
+     * @param props configuration settings
+     * @throws CopycatException
+     */
+    public abstract void start(Properties props) throws CopycatException;
```

Comment:
Hmm, maybe related to this is that it may not be clear right now where a connector developer should put code that may potentially block. Right now the way you'd do that is allocate your own thread and run something in the background (e.g. if you need to monitor for changes to the tables in a source database). Maybe we should formalize that so there is `configure(props)` and `run()` and Copycat handles allocating that thread.

However, not all connectors will need a permanently running thread (and may not need to do any checks after initial configuration), so I'm not sure requiring it for all connectors is a good idea. And I'm wondering if we should provide utilities for doing things like periodically polling.